src/shared/selinux-util.c

```diff
@@ -434,40 +444,84 @@ int mac_selinux_get_child_mls_label(int socket_fd, const char *exe, const char *
         if (r < 0)
                 return -errno;
 
-        if (!exec_label) {
+        /* If SELinuxContext= was set in the unit, use it */
+        if (exec_label) {
+                targetcon = strdup(exec_label);
+                if (!targetcon)
+                        return -ENOMEM;
+        } else {
                 /* If there is no context set for next exec let's use context
                    of target executable */
-                r = getfilecon_raw(exe, &fcon);
+                r = getfilecon_raw(exe, &targetcon);
                 if (r < 0)
                         return -errno;
         }
 
-        bcon = context_new(mycon);
-        if (!bcon)
+        /* If neither SELinuxContext= was set,
+         * nor a SELinux label exists on the executable file,
+         * use the context under which systemd is running */
+        if (!targetcon) {
+              targetcon = strdup(mycon);
+              if (!targetcon)
+                    return -ENOMEM;
+        }
+
+        tcon = context_new(targetcon);
+        if (!tcon)
                 return -ENOMEM;
 
+        /* Get context of the network socket */
         pcon = context_new(peercon);
         if (!pcon)
                 return -ENOMEM;
 
+        /* Extract last part of the network socket context
+         * (e.g.: systemd_u:object_r:netlabel_peer_t:s0 -> s0) */
         range = context_range_get(pcon);
         if (!range)
                 return -errno;
 
-        r = context_range_set(bcon, range);
+        /* Now change last part of previously computed targetcon to the
+         * one extracted from the network socket */
+        r = context_range_set(tcon, range);
         if (r)
                 return -errno;
 
-        freecon(mycon);
-        mycon = strdup(context_str(bcon));
-        if (!mycon)
-                return -ENOMEM;
-
         sclass = string_to_security_class("process");
         if (sclass == 0)
                 return -ENOSYS;
 
-        return RET_NERRNO(security_compute_create_raw(mycon, fcon, sclass, label));
+        resultcon = strdup(context_str(tcon));
+        if (!resultcon)
+                return -ENOMEM;
+
+        /* If SELinuxContext= was set explicitly, use it, replace MLS level
+         * with the one from socket and fail later if SELinux does not allow
+         * to transition to this context */
+        if (exec_label) {
+                *label = strdup(resultcon);
+                if (!*label)
+                      return -ENOMEM;
+        /* If heurestics were used, ask the Linux kernel which context it "recommends"
+         * (going here with explicitly set SELinuxContext= may result
+         * in its value silently being not applied) */
+        } else {
+                /* bcon is our current context, but with MLS level from socket;
```

Comment:
Also I guess I missing the entire point of this else branch? Isn't this case already handed above by
```C
 targetcon = strdup(mycon);
```